README.md
# mautic_tracking_api

 App [User Monitoring and Tracking](https://www.mautic.org/docs/en/contacts/contact_monitoring.html) easily for Dart and Flutter using [Mautic: Open Source Marketing Automation Software](https://github.com/mautic/mautic).

The essence of monitoring what happens in an App is similar to monitoring what happens on a website.

In short, this package use **Named Screen Path** (e.g. `main_screen`) in your App as your `page_url` field in the tracker, **Named Screen Title** (e.g. `Home Screen`) as your `page_title` field and the **Contact E-mail** as unique identifier. See [Mautic Contact Monitoring](#mautic-contact-monitoring) section for detailed instructions.

This package also use native **Mautic Tracking Cookies** `mtc_id`, `mtc_sid` and `mautic_device_id` to make Tracking more effective.

Made with :heart: by **Mautic Specialists** at [Powertic](https://powertic.com).

## Install

Add this package to your package's ```pubspec.yaml``` file:

```yaml
dependencies:
  mautic_tracking_api:
```

Run `flutter pub get` on terminal to download and install packages and import on your `main.dart`:

```dart
import 'package:mautic_tracking_api/mautic_tracking_api.dart';
```

## Usage

Create a global instance of `MauticTracking` and import this global instance on all your files.

```dart
var MauticTracking mautic;
```

On your `main.dart` you can [setting up](#setting-up) the `MauticTracking` object:

```dart
// Setting Up
void main() async {

  // Start Tracking
  mautic = MauticTracking(
    "https://mkt.mymautic.com",
    appName: "MyApp",
    email: "contact@email.com",
    appVersion: '1.0.0',
    appBundleName: 'com.mydomain.myapp',
  );

  // Track App Start
  await mautic.trackAppStart();

}
```

When the App is Started, this event will be fired:

![Started App](https://github.com/powerticmkt/dart-mautic-tracking-api/raw/master/docs/started_app.png)

If the contact doesn't exists on Mautic, the contact will be identified by `app_started`.

![Contact Identified](https://github.com/powerticmkt/dart-mautic-tracking-api/raw/master/docs/contact_identified.png)

See a complete Flutter example at [example/example_app/lib/main.dart](https://github.com/powerticmkt/dart-mautic-tracking-api/blob/master/example/example_app/lib/main.dart)

## Setting Up

There are some options to instantiate `MauticTracking`:

```dart
MauticTracking(
    this._base_url, {
    this.email,
    this.appName,
    this.appVersion,
    this.appBundleName,
    this.closeConnectionAfterRequest = true,
    this.userAgent =
        'Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.0; Trident/5.0)',
  });
```

These are the constructor options:

| Property                      | Type     | Required | Description                    |
| ----------------------------- | -------- | -------- | ------------------------------ |
| `_base_url`                   | `String` | Yes      | Mautic Base URL                |
| `email`                       | `String` | No       | Mautic Contact Email           |
| `appName`                     | `String` | No       | App Name                       |
| `appVersion`                  | `String` | No       | App Version                    |
| `appBundleName`               | `String` | No       | App Bundle Name                |
| `closeConnectionAfterRequest` | `bool`   | No       | Close Connection after Request |
| `userAgent`                   | `String` | No       | Request User Agent             |

## Mautic Contact Monitoring

The act of monitoring the traffic and activity of contacts can sometimes be somewhat technical and frustrating to understand. Mautic Tracking API makes this monitoring simple and easy to configure.

![Contact Timeline](https://github.com/powerticmkt/dart-mautic-tracking-api/raw/master/docs/contact_timeline.png)

## Tracking Screens

Send App Screen info to Mautic.

![Screen Track](https://github.com/powerticmkt/dart-mautic-tracking-api/raw/master/docs/screen_track.png)

**Method Definition:**

```dart
Future<void> trackScreen(String screenPath, [String screenName]) async;
```

| Property     | Type     | Required | Description               |
| ------------ | -------- | -------- | ------------------------- |
| `screenPath` | `String` | Yes      | Unique Screen Route Path  |
| `screenName` | `String` | No       | Unique Screen Route Label |

**Usage:**

You can send screen tracking info only with `screenPath`.

Create your screen paths using **_** to separate words and always use **small caps**.

```dart
mautic.trackScreen("main");
```

You can also send screen tracking info using `screenPath` and `screenName`. Fell free to use space and capitalized words for Screen Name.

```dart
mautic.trackScreen("main", "Main Page");
```

## Tracking Events

Send App Event info to Mautic.

![Event Track](https://github.com/powerticmkt/dart-mautic-tracking-api/raw/master/docs/event_track.png)

**Method Definition:**

```dart
Future<void> trackEvent(String eventKey, String eventName, String screenPath, [String screenName]) async;
```

| Property     | Type     | Required | Description               |
| ------------ | -------- | -------- | ------------------------- |
| `eventKey`   | `String` | Yes      | Unique Event Key          |
| `eventName`  | `String` | Yes      | Unique Event Name         |
| `screenPath` | `String` | Yes      | Unique Screen Route Path  |
| `screenName` | `String` | No       | Unique Screen Route Label |

**Usage:**

You can send screen tracking info only with `screenPath`.

Create your screen paths using **_** to separate words and always use **small caps**.

```dart
mautic.trackEvent('click_getting_start', 'Click Get Start Button', 'home');
```

You can also send screen tracking info using `screenPath` and `screenName`. Fell free to use space and capitalized words for Scren Name and Event Name.

```dart
mautic.trackEvent('click_getting_start', 'Click Get Start Button', 'home', 'Home Page');
```

### Adding Contact Tags

Add tag to contact.

![Contact Timeline](https://github.com/powerticmkt/dart-mautic-tracking-api/raw/master/docs/added_tag.png)

```dart
mautic.addTag({'a', 'b'});
```

| Property | Type  | Required | Description        |
| -------- | ----- | -------- | ------------------ |
| `params` | `Set` | Yes      | `Set` of tag names |

### Removing Contact Tags

Remove tag to contact.

![Contact Timeline](https://github.com/powerticmkt/dart-mautic-tracking-api/raw/master/docs/removed_tag.png)

```dart
mautic.removeTag({'a', 'b'});
```

| Property | Type  | Required | Description        |
| -------- | ----- | -------- | ------------------ |
| `params` | `Set` | Yes      | `Set` of tag names |

## Contribute

Please file feature requests and bugs at the [issue tracker](https://github.com/powerticmkt/dart-mautic-tracking-api/issues).


CHANGELOG.md
# Changelog

## 1.0.2

- Updated Readme

## 1.0.1

- Updated Readme
- Fix User Agent
- Flutter Example

## 1.0.0

- Initial Version
- Track App Screens
- Track App Events
- Send Cookies to Tracking
- Track using E-mail
- Add Tags
- Remove Tags
- Track App Version
- Track App Name
- Customize Tracking User Agent


example/example_app/lib/main.dart
import 'package:flutter/material.dart';
import 'package:mautic_tracking_api/mautic_tracking_api.dart';

MauticTracking mautic;

void main() async {
  ///
  runApp(MyApp());

  /// Start Tracking
  mautic = MauticTracking(
    "https://mkt.powertic.com",
    appName: "MyApp",
    email: "contact@email.com",
    appVersion: '1.0.0',
    appBundleName: 'com.mydomain.myapp',
  );

  ///
  await mautic.trackAppStart();
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    // Send Click Event
    mautic.trackEvent('click_counter', 'Click Counter Button', 'main');
    mautic.addTag({'tag1'});
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.display1,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/src/mautic_tracking.dart
library mautic_tracking_api;

import 'dart:io';

/// API to User Monitoring and Tracking using Mautic
class MauticTracking {
  /// Constructor
  MauticTracking(
    this._base_url, {
    this.email,
    this.appName,
    this.appVersion,
    this.appBundleName,
    this.closeConnectionAfterRequest = true,
    this.userAgent =
        'Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.0; Trident/5.0)',
  });

  /// Mautic Base URL
  final String _base_url;

  /// Mautic Contact Email
  final String email;

  /// App Name
  final String appName;

  /// App Version
  final String appVersion;

  /// App Bundle Name
  final String appBundleName;

  /// Close Connection to Tracking after Request?
  /// Use [true] to Close Connection or [false] to make connection opened beetween requests
  final bool closeConnectionAfterRequest;

  /// Request User Agent
  final String userAgent;

  /// Mautic Contact Cookie
  Cookie contactCookie = Cookie('mtc_id', '');

  /// Mautic Session Cookie
  Cookie sessionCookie = Cookie('mtc_sid', '');

  /// Mautic Device Cookie
  Cookie deviceCookie = Cookie('mtc_device_id', '');

  /// Mautic Tracking gif URL
  String get _tracking_url => parseURL(_base_url);

  /// Parse a URL [_url] and return authority domain
  String parseURL(String _url) {
    return _url
        .replaceAllMapped(RegExp('http(s)?:\/\/'), (match) {
          return '';
        })
        .replaceAll('/', '')
        .toLowerCase()
        .trim();
  }

  /// Return the App Name to Display on Contact Timeline
  String _getAppName() {
    return ((appName == null)
        ? ((appVersion == null) ? 'App' : 'App ($appVersion): ')
        : ((appVersion == null) ? '$appName ' : '$appName ($appVersion): '));
  }

  /// Make Request to Tracking
  Future<void> _makeRequest({Map<String, String> params}) async {
    if (email != null) {
      params.addEntries({MapEntry('email', email)});
    }

    if (appBundleName != null) {
      params.addEntries({MapEntry('page_referrer', appBundleName)});
    }

    var _uri = Uri.http(_tracking_url, 'mtracking.gif', params);

    var client = HttpClient();
    client.userAgent = userAgent;

    await client.getUrl(_uri).then((HttpClientRequest request) {
      /// Prepare Request
      request.headers.set('Accept-Language', 'en-us');
      request.headers.set('Accept-Encoding', 'gzip, deflate');

      /// Setting Up Cookies
      if (contactCookie.value.isNotEmpty) {
        request.cookies.add(contactCookie);
        request.cookies.add(deviceCookie);
        request.cookies.add(sessionCookie);
      }

      /// Make Request
      return request.close();
    }).then((HttpClientResponse response) {
      /// Return Response
      for (var i = 0; i < response.cookies.length; i++) {
        var _cookie = response.cookies[i];
        if (_cookie.name.contains('mtc_sid')) {
          sessionCookie.value = _cookie.value;
          sessionCookie.httpOnly = false;
        }
        if (_cookie.name.contains('mautic_device_id')) {
          deviceCookie.value = _cookie.value;
          deviceCookie.httpOnly = false;
        }
        if (_cookie.name.contains('mtc_id')) {
          contactCookie.value = _cookie.value;
          contactCookie.httpOnly = false;
        }
      }
    });
    client.close(force: closeConnectionAfterRequest);
  }

  /// Track the App Start
  ///
  /// Example:
  ///
  /// ```dart
  /// // Track the App Start
  /// await trackAppStart();
  /// ```
  Future<void> trackAppStart() async {
    await _makeRequest(
      params: {
        'page_url': 'app_started',
        'page_title': '${_getAppName()} Started',
      },
    );
  }

  /// Send App Screen info to Tracking
  ///
  /// Use [screenPath] to register your unique **Screen Route Path** and [screenName] as **Screen Label**.
  ///
  /// Examples:
  ///
  /// ```dart
  /// // Send only Screem Path
  /// trackScreen('dashboard');
  /// ```
  /// ```dart
  /// // Send Screen Path and Timeline Name
  /// trackScreen('view_contact', 'View Contact Info');
  /// ```
  Future<void> trackScreen(String screenPath, [String screenName]) async {
    if (screenName == null) {
      await _makeRequest(
        params: {
          'page_url': screenPath,
          'page_title': '${_getAppName()} Navigation',
        },
      );
    } else {
      await _makeRequest(
        params: {
          'page_url': screenPath,
          'page_title': '${_getAppName()} Navigation: $screenName',
        },
      );
    }
  }

  /// Send App Event info to Tracking
  ///
  /// Use [eventKey] to register your unique Event Key and [eventName] as Event Label,
  /// and [screenPath] to register your Routes and [screenName] as Label.
  ///
  /// Examples:
  ///
  /// ```dart
  /// // Sent Event Key, Event Label and Screen Path
  /// trackEvent('click_total', 'Click Dashboard Total Button', 'dashboard');
  /// ```
  /// ```dart
  /// // Sent Event Key, Evebt Label, Screen Path and Screen Name
  /// trackEvent('change_password', 'Change User Password', 'user_info', 'User Info');
  /// ```
  Future<void> trackEvent(
    String eventKey,
    String eventName,
    String screenPath, [
    String screenName,
  ]) async {
    if (screenName == null) {
      await _makeRequest(
        params: {
          'page_url': 'screen_${screenPath}_event_${eventKey}',
          'page_title': '${_getAppName()} Event: $eventName'
        },
      );
    } else {
      await _makeRequest(
        params: {
          'page_url': 'screen_${screenPath}_event_${eventKey}',
          'page_title': '${_getAppName()} Event: $screenName / $eventName'
        },
      );
    }
  }

  /// Modify Contact Tags
  Future<void> _changeTag(Set<String> _tags, {bool addOperation = true}) async {
    _tags = _tags
        .map(
          (item) => (addOperation) ? item : '-$item',
        )
        .toSet();
    await _makeRequest(params: {
      'page_title': ('${_getAppName()} ' +
          (addOperation ? 'Added Tags: ' : 'Removed Tags: ') +
          _tags.join(',')),
      'tags': _tags.join(',')
    });
  }

  /// Add Contact Tags
  ///
  /// Examples:
  ///
  /// ```dart
  /// // Sent Event Key, Event Label and Screen Path
  /// addTag({'tag1', 'tag2'});
  /// ```
  Future<void> addTag(Set<String> _tags) async {
    await _changeTag(_tags);
  }

  /// Remove Contact Tags
  ///
  /// Examples:
  ///
  /// ```dart
  /// // Sent Event Key, Event Label and Screen Path
  /// removeTag({'tag1', 'tag2'});
  /// ```
  Future<void> removeTag(Set<String> _tags) async {
    await _changeTag(_tags, addOperation: false);
  }
}
